### ISSUE TITLE
JTReg Test Failure: java/lang/String/StringRepeat.java

### ISSUE THREAD

**Issue by adam-thorpe** [eclipse/openj9#6667]
Failure link
------------
https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/openj9/test/jdk/java/lang/String/StringRepeat.java
Test Category: openjdk
Test Suite Name: jdk_lang
OS/Architecture: x86-64-windows, x86-64-linux, x86-64-osx, ...
OpenJDK Version: 11.0.4+11
Public Build SHAs:
  OpenJ9   - `1fcc869f3`
  OMR        - `ee5b41ac`
  JCL          - `965d0c0df3`

Optional info
-------------
Test fails with jdk11 openj9, passes on hotspot. Issue seems to be with `String.repeat()` where passing in the value `Integer.MAX_VALUE - 1` does not throw `OutOfMemoryError` as intended but instead throws `NegativeArraySizeException`.

Failure output
--------------
```
java.lang.NegativeArraySizeException
        at java.base/java.lang.String.repeat(String.java:3919)
        at StringRepeat.test2(StringRepeat.java:85)
        at StringRepeat.main(StringRepeat.java:35)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
        at java.base/java.lang.Thread.run(Thread.java:831)
```

**Comment by smlambert** [comment]
@adam-thorpe - for test failure issues, the failure link would be the link to the job where the test fails (though it is also good that you include the link to the test case itself) (which will tell us which JDK_VERSION, OS/arch, test suite name, OpenJDK version).  

Ideally, as per the [issue template ](https://github.com/eclipse/openj9/blob/master/.github/ISSUE_TEMPLATE/test-suite-failure.md), we extract that info from the test link itself (since Jenkins jobs do not stay around for long, and we need to capture that info, in case no one gets to the issue before the link is no longer valid): 

test category (functionality, system, etc): openjdk
test suite name: jdk_lang
OS/architecture (i.e. Ubuntu 16.04 x64): ? what platforms is this test failing on?
OpenJDK version (i.e. 8u202-b08): 
public build SHAs (i.e. -version output):  in the console output of the failing test, find the java -version output

This info is needed so that developers can rerun the test and reproduce the failure (think about what you need to run a Grinder job):
Test category -> (BUILD_LIST=openjdk)
test suite name -> (TARGET=jdk_lang, or jdk_custom with CUSTOM_TARGET=test/jdk/java/lang/String/StringRepeat.java)
OS/arch -> Jenkinsfile

<img width="639" alt="Screen Shot 2019-08-05 at 9 18 06 AM" src="https://user-images.githubusercontent.com/2836948/62467521-f87a8300-b761-11e9-9c8d-0cd1dfe176d2.png">

**Comment by adam-thorpe** [comment]
For some of these tests, I did locally so there is no jenkins link. Would you like me to create a jenkins build and then run that or just manually input the information here that you would use to create the job?

**Comment by smlambert** [comment]
You do not necessarily need to create a Jenkins job recreating the issue (though it might be interesting to note if there are other platforms that it fails on besides your laptop, which is likely either osx or windows).

The issue template says, 

"If the link is not public, instead include

test category (functionality, system, etc):
test suite name:
OS/architecture (i.e. Ubuntu 16.04 x64):
OpenJDK version (i.e. 8u202-b08):
public build SHAs (i.e. -version output):

Essentially you need to provide enough info that the openj9 developers know how to run the test and reproduce the failure?  including what platform(s) you see the failure on

**Comment by andrew-m-leonard** [comment]
just had a quick look at the openj9 String.repeat(count) method and it looks as though the method is getting a NegativeArraySizeException when using a calculated total repeated string length and creating an array using it. Due to the large count value the int repeat length overflows to a negative value which is used in the array... hence exception...

**Comment by adam-thorpe** [comment]
https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/openjdk/ProblemList_openjdk11-openj9.txt#L57

**Comment by darshitpp** [comment]
I am trying to reproduce this problem by running the tests, but I seem to be missing something. I have already read https://github.com/eclipse/openj9/blob/master/test/README.md and https://github.com/eclipse/openj9/blob/master/test/docs/OpenJ9TestUserGuide.md and I am accordingly trying to run with the following in the `~/openj9-openjdk-jdk11/openj9/test/TKG` directory

```
export TEST_JDK_HOME=<path to my local OpenJDK 11>
export BUILD_LIST=openjdk
export TARGET=jdk_lang
make compile
```
This works. However, neither of the commands `make _openjdk` or `make _jdk_lang` runs the test for me. Where in the steps am I going wrong?

**Comment by pshipton** [comment]
`export JDK_VERSION=11` is missing. 
`make _jdk_lang_j9_0` should work.

**Comment by pshipton** [comment]
`make _jdk_custom JDK_CUSTOM_TARGET=test/jdk/java/lang/String/StringRepeat.java` should also work to run just the failing test.

**Comment by darshitpp** [comment]
> `export JDK_VERSION=11` is missing.

It did auto-detect the JDK version, so I did not mention that.

> `make _jdk_lang_j9_0` should work.

Can you explain to me how to arrive at `_jdk_lang_j9_0` since the docs don't mention anything like that?

Even the command for `JDK_CUSTOM_TARGET` only has a fleeting reference in the docs and I wouldn't even have managed to figure this out.

However, the above still do not work.

The output for `make _jdk_lang_j9_0` says
```
Warning: cannot find the following tests [jdk_lang_j9_0] in TESTLIST
.
.
There is no test for target jdk_lang_j9_0
```
I did even try to put the test in the test list using `export TESTLIST=jdk_lang_j9_0` but it had the same result

**Comment by pshipton** [comment]
I didn't set `export TARGET=jdk_lang`, and running `make _jdk_lang_j9_0` worked for me.

I figured out the target name by looking at the OpenJ9 nightly build, i.e. https://ci.eclipse.org/openj9/job/Test_openjdk11_j9_sanity.openjdk_ppc64le_linux_Nightly/106/tapTestReport/

You can also see the targets in the playlist https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/openjdk/playlist.xml#L241

The ones with `<impl>openj9</impl>` apply to OpenJ9.

**Comment by smlambert** [comment]
re: https://github.com/eclipse/openj9/issues/6667#issuecomment-655293122 it appears the test docs were not updated after the test target names were changed via [openjdk-tests/1699](https://github.com/AdoptOpenJDK/openjdk-tests/pull/1699).  

Assume documentation is always stale.

test targets are defined in playlist.xml files, playlists contain the 'truth' about test targets more than documentation does

test targets that belong to the openjdk test group are defined/listed in https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/openjdk/playlist.xml. openjdk tests can be excluded via ProblemLists files (a.k.a. excludes files) also found in the [openjdk-tests/openjdk directory](https://github.com/AdoptOpenJDK/openjdk-tests/tree/master/openjdk).

[jdk_lang_j9](https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/openjdk/playlist.xml#L241-L264) is a test target that encapsulates all the tests in the java/lang package as defined in a mirror or [extensions repo of openjdk](https://github.com/ibmruntimes/openj9-openjdk-jdk11).  Any tests that have been added to the ProblemList file will not be run.  StringRepeat was excluded because it is failing, see https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/openjdk/ProblemList_openjdk11-openj9.txt#L57, also noting the "test excluded" label on this issue.

jdk_custom is a test target that allows you to run a single test class, but you need to also tell the test framework the name of the test class by exporting [JDK_CUSTOM_TARGET](https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/openjdk/playlist.xml#L47).  It will ignore the ProblemList file and not run the other tests in the java/lang package, so it is your best choice to rerun this test.  

Alternatively you could create a branch of openjdk-tests and remove StringRepeat from the ProblemList file and run _jdk_lang_j9 which would then include that un-excluded test class, but this is less approachable than using jdk_custom

**Comment by pshipton** [comment]
This works for me
```
export TEST_JDK_HOME=<path to my local OpenJDK 11>
export BUILD_LIST=openjdk
git clone git@github.com:AdoptOpenJDK/openjdk-tests.git
cd openjdk-tests/
git clone git@github.com:AdoptOpenJDK/TKG.git
cd TKG/
make compile
make _jdk_custom JDK_CUSTOM_TARGET=test/jdk/java/lang/String/StringRepeat.java
```

**Comment by pshipton** [comment]
and then see `test_output_15942181262469/jdk_custom_0/work/java/lang/String/StringRepeat.jtr`

**Comment by smlambert** [comment]
<id> is the timestamp for when that test execution was started

**Comment by darshitpp** [comment]
The [Quick Start Guide for running tests](https://github.com/eclipse/openj9/blob/master/test/README.md) has no mention of cloning the `openjdk-tests` repository, and it only mentions the `TKG` repository. As both of your comments mention the `openjdk-tests` repository, I would clone it, and retry. However, I am now confused about how I managed to run the test suite using just `TKG`. 

I want to confirm if I shall fork the code to my profile and then clone the main openj9 repo (and just clone but not fork the test/TKG/omr repositories). 

If the Quick Start Guide is indeed incomplete, could I attempt fixing it too?

**Comment by smlambert** [comment]
I have an [open issue to centralize the test documentation](https://github.com/eclipse/openj9/issues/6318).  Feel free to attempt to fix it, but it is related of a major overhaul of test doc, started here: https://github.com/AdoptOpenJDK/openjdk-tests/pull/1556

The Quick Start Guide was written to help openj9 developers in an assumed common user story, they are working in their openj9 branch locally and wanting to run **functional** tests which also live in openj9 repo.  So it should really be named Quick Start Guide for running functional tests.  Since you want to run **openjdk** tests, the Quick Start Guide does not contain the instructions for that.  

TKG looks at what your BUILD_LIST is, and fetches necessary test material to accomplish the execution of that material, so you do not have to explicitly fetch it yourself.  (TKG always fetches openjdk-tests, and if BUILD_LIST contains openjdk, it additionally fetches the jdk extensions repo.  This is done implicitly so people do not have to go fetching resources themselves).

Use this wiki page as a guide, if you want instructions for running any of the test groups (**functional**, **system**, **openjdk**, **perf**, **external**) : https://github.com/eclipse/openj9/wiki/Reproducing-Test-Failures-Locally

**Comment by pshipton** [comment]
Doing https://github.com/eclipse/openj9/issues/6667#issuecomment-655552131 without cloning the openjdk-tests didn't work for me. The `make compile` didn't do much, and trying to run the test shows `Warning: cannot find the following tests [jdk_custom] in TESTLIST`

**Comment by smlambert** [comment]
I was commenting on the [quick start guide content](https://github.com/eclipse/openj9/blob/master/test/README.md#quick-start) which says clone **TKG repo**.  If you follow those instructions it will **also** work, but it confuses people so it will be changed to clone the **openjdk-tests repo**.  Cloning either works (for functional tests), but I will do my very best to fix our doc so as not confuse the dev team.

**Comment by darshitpp** [comment]
`StringRepeat` has two tests - `test1()` and `test2()`, but the test execution only shows that one test was executed, and passed. There is no failed test. I ran the test on the Jenkins Grinder job (https://ci.adoptopenjdk.net/job/Grinder/3558/) and received the same results.

Does this mean the String repeat is a single test (or showing consolidated results), or is there something missing which is not making the 2nd test fail? (Or perhaps the failed test case has already been fixed)
